spec/app/domain/authentication/validate_security_spec.rb

```diff
@@ -251,4 +244,52 @@ def resource_class(returned_resource)
       expect { subject }.to raise_error(Errors::Authentication::Security::AccountNotDefined)
     end
   end
+
+  context "when validating the same role a second time" do
+    let(:user_id) { 'some-user' }
+    let(:user_roleid) { [test_account, 'user', user_id].join(':') }
+    let(:user_role_double) { double('user_role_double') }
+
+    # We can't use the double returned by the +role_class+ function
+    # above, because it doesn't constrain the arguments to +:[]+.
+    let(:role_class) { double('role_class') }
+
+    before do
+      allow(role_class).to receive(:roleid_from_username).and_return(user_roleid)
+      allow(role_class).to receive(:[])
+        .with("#{test_account}:user:admin")
+        .and_return(user_role(is_authorized: true))
+      allow(user_role_double).to receive(:allowed_to?).and_return(true)
+    end
+
+    subject do
+      described_class
+        .new(
+          role_class: role_class,
+          webservice_resource_class: full_access_resource_class
+        ).(
+          webservice: webservice('service1'),
+          account: test_account,
+          user_id: user_id,
+          enabled_authenticators: two_authenticator_env
+        )
+    end
+    
+    it "role lookups are not cached" do
+
+      # Simulate two validations. For the first, the role should not
+      # be found.
+      allow(role_class).to receive(:[]).with(user_roleid).and_return(nil)
+      expect { subject }.to raise_error(Errors::Authentication::Security::UserNotDefinedInConjur)
+
+      # For the second, the role should be found, and validation
+      # should succeed.
+
+      # Note that, because the arguments are the same, this +allow+
+      # overwrites the previous one.
+      allow(role_class).to receive(:[]).with(user_roleid).and_return(user_role_double)
+      expect { subject }.not_to raise_error
```

Comment:
I believe this is not testing what we want it to test.  `subject` will recreate a new instance each time it's called, whereas we want a single subject with the behaivor of `role_class` changing.  

Instead you'll need to create an instance of `described_class` and assign it to a local variable at the top of this `it` block.  Then explicitly call it
```ruby
call.(
          webservice: webservice('service1'),
          account: test_account,
          user_id: user_id,
          enabled_authenticators: two_authenticator_env
        )
```
two times, chancing the `role_class` in between.  I *think* that last part will work, but you'll need to try it....

Reply:
No, that's not right. The original bug was that *two different instances* of the validator were sharing the same memoization of a `Role` lookup (one that failed the first time it was attempted).

This tests that two different instances with access to the same role class implementation don't share state that way.